File: examples/play-2-scala/app/controllers/Todos.scala
package controllers

import play.api.mvc.{Result, Action, Controller}
import play.api.libs.json.Json._
import models.Todo
import data.Store

object Todos extends Controller {

  def index = Action {
    Ok(toJson(Store.getAll))
  }

  def create = TodoAction { todo =>
    implicit val writes = Todo.todoWrites
    Ok(toJson(Some(Store.save(todo))))
  }

  def update(id: Int) = TodoAction { todo =>
    implicit val writes = Todo.todoWrites
    Ok(toJson(Some(Store.save(id, todo))))
  }

  def delete(id: Int) = Action {
    Ok(toJson(Store.remove(id)))
  }

  private def TodoAction(operation: Todo => Result) = Action(parse.json) { request =>
    Todo.form.bind(request.body).fold(
      errors => BadRequest(errors.errorsAsJson),
      todo => operation(todo)
    )
  }
}

File: examples/play-2-java/app/controllers/Todos.java
package controllers;

import com.google.common.base.Optional;
import data.Store;
import models.Todo;
import org.codehaus.jackson.JsonNode;
import play.*;
import play.libs.Json;
import play.mvc.*;

import views.html.*;

public class Todos extends Controller {

    public static Store store = new Store();

    public static Result index() {
        return ok(Json.toJson(store.getAll()));
    }

    @BodyParser.Of(BodyParser.Json.class)
    public static Result create(){
        JsonNode json = request().body().asJson();
        models.Todo data = Json.fromJson(json, models.Todo.class);
        models.Todo todo = store.save(data);
        return ok(Json.toJson(todo));
    }

    @BodyParser.Of(BodyParser.Json.class)
    public static Result update(Long id) {
        JsonNode json = request().body().asJson();
        models.Todo data = Json.fromJson(json, models.Todo.class);
        Optional<models.Todo> todo = store.save(id, data);
        return ok(Json.toJson(todo));
    }

    public static Result delete(Long id) {
        Optional<models.Todo> todo = store.remove(id);
        return ok(Json.toJson(todo));
    }

}
